src/wp-includes/class-wp-scripts.php

```diff
@@ -714,6 +839,207 @@ public function in_default_dir( $src ) {
+	 */
+	public function has_delayed_inline_script() {
+		foreach ( $this->registered as $handle => $script ) {
+			foreach ( array( 'before', 'after' ) as $position ) {
+				if (
+					$this->get_data( $handle, $position ) &&
+					$this->should_delay_inline_script( $handle, $position )
+				) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Gets all dependents of a script.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $handle The script handle.
+	 * @return string[] Script handles.
+	 */
+	private function get_dependents( $handle ) {
+		// Check if dependents map for the handle in question is present. If so, use it.
+		if ( array_key_exists( $handle, $this->dependents_map ) ) {
+			return $this->dependents_map[ $handle ];
+		}
+
+		$dependents = array();
+
+		// Iterate over all registered scripts, finding dependents of the script passed to this method.
+		foreach ( $this->registered as $registered_handle => $args ) {
+			if ( in_array( $handle, $args->deps, true ) ) {
+				$dependents[] = $registered_handle;
+			}
+		}
+
+		// Add the handles dependents to the map to ease future lookups.
+		$this->dependents_map[ $handle ] = $dependents;
+
+		return $dependents;
+	}
+
+	/**
+	 * Checks if the strategy passed is a valid delayed (non-blocking) strategy.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string $strategy The strategy to check.
+	 * @return bool True if $strategy is one of the delayed strategies, otherwise false.
+	 */
+	private function is_delayed_strategy( $strategy ) {
+		return in_array(
+			$strategy,
+			$this->delayed_strategies,
+			true
+		);
+	}
+
+	/**
+	 * Checks if all of a script's dependents are delayed, which is required to maintain execution order.
+	 *
+	 * @since 6.3.0
+	 *
+	 * @param string   $handle     The script handle.
+	 * @param bool     $async_only Whether to limit to async strategy only.
+	 * @param string[] $checked    Optional. An array of already checked script handles, used to avoid recursive loops.
+	 * @return bool True if all dependents are delayed, false otherwise.
+	 */
+	private function has_only_delayed_dependents( $handle, $async_only = false, $checked = array() ) {
+		// If this node was already checked, this script can be deferred and the branch ends.
+		if ( in_array( $handle, $checked, true ) ) {
+			return true;
+		}
+
+		$checked[]  = $handle;
+		$dependents = $this->get_dependents( $handle );
+
+		// If there are no dependents remaining to consider, the script can be deferred.
+		if ( empty( $dependents ) ) {
+			return true;
+		}
+
+		// Consider each dependent and check if it is delayed.
+		foreach ( $dependents as $dependent ) {
+			// If the dependent script has no src, ignore it for consideration.
+			if ( empty( $this->registered[ $dependent ]->src ) ) {
+				continue;
+			}
+
+			// If the dependency is not enqueued, ignore it for consideration.
+			if ( ! $this->query( $dependent, 'enqueued' ) ) {
+				continue;
+			}
```

Comment:
Just realized this may not work entirely as expected. Consider the case when a plugin is printing out script handles manually:

```php
wp_print_scripts( array( 'dependency', 'dependent' ) );
```

If `dependent` is blocking, but `dependency` is defer, then this would cause a problem. 

Not entirely sure this is is a problem yet, but it just came to mind. 


Introduced in https://github.com/10up/wordpress-develop/pull/74

Reply:
I think this is a more exhaustive regex that matches calls to `wp_print_scripts()`, `wp_scripts()->do_items()` and `$wp_scripts->do_items()` which is passed either a multi-element array _or_ a variable:

```php
(wp_print_scripts|(wp_scripts\(\)|scripts)->do_items)\(\s*(\$|((array\(|\[).+?,))
```

I see there are currently [206 results](https://wpdirectory.net/search/01H2RDES76PQSTYQMZDEGTQDPZ). 

> Regardless, there is a case to be made that if someone is manually printing scripts like this, then it's their responsibility to make sure that execution order of any inline scripts are working as expected, since these are bypassing the queue. Given the opt-in nature of loading strategies entirely, there is no backward compatibility break possible here either.

You may be entirely right here. But I do want to add some tests to check this case so we can be sure of what happens.

Reply:
> `wp_print_scripts( array( 'dependency', 'dependent' ) );`

Yea, this is a "doing_it_wrong" way to specify a dependency regardless of when it is used. Like @joemcgill I also think that "it's their responsibility to make sure that execution order of any inline scripts are working as expected". Imho the only thing core could do is to document this case of doing_it_wrong.